vllm/v1/worker/gpu_model_runner.py

```diff
@@ -888,16 +977,24 @@ def execute_model(
             scheduler_output,
         )
 
-        # Update with the actual token ids
+        # Update batch with the valid generated tokens.
+        valid_mask = sampled_token_ids != INVALID_TOKEN_ID
+        gen_lens = valid_mask.sum(dim=1).tolist()
+        valid_sampled_token_ids = [
+            seq.tolist()
+            for seq in sampled_token_ids[valid_mask].split(gen_lens)
+        ]
+        self.input_batch.num_tokens[:num_reqs] += gen_lens
         for i, req_state, seq_len in request_seq_lens:
-            token_id = sampled_token_ids[i]
-            self.input_batch.token_ids_cpu[i, seq_len] = token_id
-            req_state.output_token_ids[-1] = token_id
+            target_slice = slice(seq_len - gen_lens[i] + 1, seq_len + 1)
+            self.input_batch.token_ids_cpu[
+                i, target_slice] = valid_sampled_token_ids[i]
+            req_state.output_token_ids.extend(valid_sampled_token_ids[i])
```

Comment:
Maybe have an if/else here for simpler non-spec-decoding path?

Reply:
+1 for @njhill's comment.